### DISCUSSION TITLE
Custom problem normalizer not working in production environment

### DISCUSSION THREAD

**Discussion by tsilia** [symfony/symfony#59025] [Q&A] [answered]
Hello.
I'm using Symfony 7.0 (`and PHP 8.2.26`) in a web application that has some API which must always return JSON.

As adviced here: [Overriding Error output for non-HTML formats](https://symfony.com/doc/7.0/controller/error_pages.html#overriding-error-output-for-non-html-formats), I wrote custom problem normalizer to serve exception info as a JSON in my API controllers:
```php
//...
class JsonProblemNormalizer implements NormalizerInterface
{
    public function normalize($exception, ?string $format = null, array $context = []): array
    {
        return [
            'status' => 'error',
            'message' => $exception->getMessage(),
            'code' => $exception->getStatusCode(),
        ];
    }
    // ...
}
```

It works in dev environment: in case of an exception in my API controller I get a simple JSON.
For example, doing this:
```php
use Symfony\Component\HttpFoundation\JsonResponse;
// ...
#[Route('/api_broken_json_response_poc', name: 'api_broken_json_response_poc', methods: ['GET'], format: "json")]
public function api_broken_json_response_poc(Request $request): JsonResponse
{
    throw new RuntimeException('Unhandled runtime exception');
}
```
produces the following response:
```json
{"status":"error","message":"Unhandled runtime exception","code":500}
```
But it somehow won't work in production and in case of any exceptions instead of a JSON I get an error page with HTML.
What am I doing wrong?
Thanks in advance.

**Comment by tsilia** [comment, marked as answer]
Ok. I've found the root cause.
I'm using a custom error controller to render custom error pages using my own template:
```yaml
when@prod:
    framework:
        error_controller: App\Controller\ErrorController::show
```
```php
//...
class ErrorController extends AbstractController
{
    public function show(\Throwable $exception, ?DebugLoggerInterface $logger)
    {
        return $this->render('error.html.twig', [
                'error' => $exception->getMessage(),
            ]);
    }
}
```
which seems to interfere with error reporting and prevents exceptions from being served as JSONs when an API controller fails with an unhandled exception.
There're at least 3 possible solutions:
1. Turn off (or get rid of) the custom error controller which is kinda undesirable.
2. Modify the custom error controller to check the request and, in case of `Accept: application/json`, serve a JSON response with an instance of FlattenException which you can easily create using the `FlattenException::createFromThrowable()` method. This is the preferred solution, IMO.
3. Go messing around with kernel events (`onKernelException` and what not) which is redundant since all the functionality of serving exceptions as JSONs is already there, implemented in Symfony.

I went with the 2nd.

If anyone has better ideas, please share! Thanks in advance.
I'll mark this as solved soon anyway.

**Reply by alexandre-le-borgne** [reply]
```
framework:
    handle_all_throwables: true
```
    
https://symfony.com/doc/current/reference/configuration/framework.html#handle-all-throwables


You may need to create that subscriber (It is possible that a newer version of Symfony does this automatically but I am not sure) (test without it before adding this and adapting it to your code) :

```
readonly class ExceptionSubscriber implements EventSubscriberInterface
{
    public function __construct(private JsonProblemNormalizer $normalizer)
    {
    }

    #[\Override]
    public static function getSubscribedEvents(): array
    {
        return [
            KernelEvents::EXCEPTION => [
                ['onException', 0],
            ],
        ];
    }

    public function onException(ExceptionEvent $event): void
    {
        $exception = $event->getThrowable();

        if ($this->normalizer->supportsNormalization($exception)) {
            /**
             * @var string|array $payload
             */
            ['payload' => $payload, 'statusCode' => $statusCode] = $this->normalizer->normalize($exception);
            $event->setResponse($new JsonResponse($payload, $statusCode));
        }
    }
}

```

**Comment by tsilia** [comment]
Thanks for your input, but as I said I would like to avoid messing around with kernel events.
As for `handle_all_throwables`, I already have it set to true.

**Reply by alexandre-le-borgne** [reply]
This is now supposed to be handled natively by symfony (https://symfony.com/doc/current/controller/error_pages.html#overriding-error-output-for-non-html-formats).

Check that you are sending the Accept: application/json header to your API and that it is receiving it correctly.

You can try debugging the automatic serialization here (in particular that the format is json and not plain/text or html): 

https://github.com/symfony/symfony/blob/67c835ee47b7fe1042a2b9a849e851930b325339/src/Symfony/Component/Serializer/Normalizer/ProblemNormalizer.php#L59
